Makefile

```diff
@@ -175,70 +167,123 @@ tidy: clean
 	GO111MODULE=on go mod tidy
 	git diff -U --exit-code go.mod go.sum
 
-update-install-scipt:
+install.sh:
 	./hack/update_install_script.sh
 
-check-install-script: update-install-scipt
+check-install-script: install.sh
 	git diff -U --exit-code install.sh
 
 clean:
-	rm -rf docs/docs.go
+	rm -rf docs/docs.go $(CLEAN_TARGETS)
 
 boilerplate:
 	./hack/verify-boilerplate.sh
 
-taily-build:
-	if [ "$(shell docker ps --filter=name=$@ -q)" = "" ]; then \
-		docker build -t pingcap/chaos-binary ${DOCKER_BUILD_ARGS} .; \
-		docker run --rm --mount type=bind,source=$(shell pwd),target=/src \
-			--name $@ -d pingcap/chaos-binary tail -f /dev/null; \
-	fi;
-
-taily-build-clean:
-	docker kill taily-build && docker rm taily-build || exit 0
+image: image-chaos-daemon image-chaos-mesh image-chaos-dashboard
 
-image: image-chaos-daemon image-chaos-mesh image-chaos-dashboard image-chaos-fs image-chaos-scripts
+GO_TARGET_PHONY :=
 
-image-chaos-mesh-protoc:
-	docker build -t pingcap/chaos-mesh-protoc ${DOCKER_BUILD_ARGS} ./hack/protoc
+define COMPILE_GO_TEMPLATE =
+ifeq ($(CHAOS_MESH_BUILD_IN_DOCKER),0)
 
-ifneq ($(TAILY_BUILD),)
-image-binary: taily-build image-build-base
-	docker exec -it taily-build make binary
-	cp -r scripts ./bin
-	echo -e "FROM scratch\n COPY . /src/bin\n COPY ./scripts /src/scripts" | docker build -t pingcap/chaos-binary -f - ./bin
+$(1): $(4)
+ifeq ($(3),1)
+	$(CGO) build -ldflags "$(LDFLAGS)" -tags "${BUILD_TAGS}" -o $(1) $(2)
 else
-image-binary: image-build-base
-	DOCKER_BUILDKIT=1 docker build -t pingcap/chaos-binary ${DOCKER_BUILD_ARGS} .
+	$(GO) build -ldflags "$(LDFLAGS)" -tags "${BUILD_TAGS}" -o $(1) $(2)
+endif
+
+endif
+GO_TARGET_PHONY += $(1)
+endef
+
+BUILD_INDOCKER_ARG := --env IN_DOCKER=1
+ifneq ($(GO_BUILD_CACHE),)
+	BUILD_INDOCKER_ARG += --volume $(GO_BUILD_CACHE)/chaos-mesh-gopath:/tmp/go
+	BUILD_INDOCKER_ARG += --volume $(GO_BUILD_CACHE)/chaos-mesh-gobuild:/tmp/go-build
+endif
+define BUILD_IN_DOCKER_TEMPLATE =
+CLEAN_TARGETS += $(2)
+ifeq ($(CHAOS_MESH_BUILD_IN_DOCKER),1)
+$(2): image-build-env go_build_cache_directory
+	docker run --rm --workdir /mnt/ --volume $(shell pwd):/mnt \
+		$(BUILD_INDOCKER_ARG) --env UI=${UI} --env SWAGGER=${SWAGGER} \
+		--user $(shell id -u):$(shell id -g) ${DOCKER_REGISTRY_PREFIX}pingcap/build-env \
+		/usr/bin/make $(2)
+endif
+image-$(1)-dependencies := $(image-$(1)-dependencies) $(2)
+endef
```

Comment:
do we need to add some indent in this block?

Reply:
Makefile is indent-sensitive. I don't know whether I can add indents without breaking it.